tests/test_auth.py

```diff
@@ -0,0 +1,369 @@
+
+    envoy = Envoy("127.0.0.1")
+    await envoy.setup()
+    with pytest.raises(EnvoyAuthenticationError):
+        await envoy.authenticate("username", "password")
+
+
+@pytest.mark.asyncio
+@respx.mock
+async def test_no_remote_token_with_7_6_175_standard() -> None:
+    """Test Unable to obtain token for Envoy authentication from https://entrez.enphaseenergy.com/tokens"""
+    logging.getLogger("pyenphase").setLevel(logging.DEBUG)
+    version = "7.6.175_standard"
+    start_7_firmware_mock()
+    prep_envoy(version, info=True)
+
+    respx.post("https://enlighten.enphaseenergy.com/login/login.json?").mock(
+        return_value=Response(
+            200,
+            json={
+                "session_id": "1234567890",
+                "user_id": "1234567890",
+                "user_name": "test",
+                "first_name": "Test",
+                "is_consumer": True,
+                "manager_token": "1234567890",
+            },
+        )
+    )
+    respx.post("https://entrez.enphaseenergy.com/tokens").mock(
+        return_value=Response(500, text="token")
+    )
+    respx.get("/auth/check_jwt").mock(return_value=Response(200, json={}))
+
+    envoy = Envoy("127.0.0.1")
+    await envoy.setup()
+    with pytest.raises(EnvoyAuthenticationError):
+        await envoy.authenticate("username", "password")
+
+
+@pytest.mark.asyncio
+@respx.mock
+async def test_enlighten_json_error_with_7_6_175_standard() -> None:
+    """Test Unable to decode response from Enlighten"""
+    logging.getLogger("pyenphase").setLevel(logging.DEBUG)
+    version = "7.6.175_standard"
+    start_7_firmware_mock()
+    prep_envoy(version, info=True)
+
+    respx.post("https://enlighten.enphaseenergy.com/login/login.json?").mock(
+        return_value=Response(
+            200,
+            text="nojson",
+        )
+    )
+    respx.post("https://entrez.enphaseenergy.com/tokens").mock(
+        return_value=Response(200, text="token")
+    )
+    respx.get("/auth/check_jwt").mock(return_value=Response(200, json={}))
+
+    envoy = Envoy("127.0.0.1")
+    await envoy.setup()
+    with pytest.raises(EnvoyAuthenticationError):
+        await envoy.authenticate("username", "password")
+
+
+@pytest.mark.asyncio
+@respx.mock
+async def test_token_with_7_6_175_standard() -> None:
+    """Test auth using token"""
+    logging.getLogger("pyenphase").setLevel(logging.DEBUG)
+    version = "7.6.175_standard"
+    start_7_firmware_mock()
+    prep_envoy(version, info=True)
+
+    envoy = Envoy("127.0.0.1")
+    await envoy.setup()
+
+    token = jwt.encode(
+        payload={"name": "envoy", "exp": 1707837780}, key="secret", algorithm="HS256"
+    )
+
+    await envoy.authenticate("username", "password", token)
+    assert isinstance(envoy.auth, EnvoyTokenAuth)
+    assert envoy.auth.expire_timestamp == 1707837780
+    assert envoy.auth.token == token
+    # FIXME # next 2 reveal issue in auth.py
```

Comment:
It would be better to split this and mark it xfail

Reply:
Think this is a problem in auth that surfaced while building the tests. That's why the comment is there and I left it in. self._is_customer and self._manager_token is not defined at the EnvoyTokenAuthclass level, only used and assigned in the _obtain_token method. 

Both EnvoyTokenAuthclass manager_token property and is_customer property throw an error that the private variable doesn't exist on the class level. 

Since these are only filled when connecting to Enlighten, one can ask if they are really needed, the will hardly exist.